### ISSUE TITLE
Scope HTTPS client certificate requirement to specific paths in Kestrel

### ISSUE THREAD

**Issue by leastprivilege** [aspnet/AspNetCore#18064]
### Is your feature request related to a problem? Please describe.
I am trying to run IdentityServer4 with the OAuth 2.0 MTLS [spec](https://tools.ietf.org/html/draft-ietf-oauth-mtls-17) on a Kestrel host.

The spec requires that certain endpoints must require HTTPS - in our case we put them all under a common path, e.g. /connect/mtls/*. For IIS deployments we use the location element in web.config to set the client cert config for that path.

In Kestrel there is only a global client cert ignore/allow/require switch - when I set this to `allow` (as described [here](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.1)) it will affect the whole application, e.g. when browsing to a completely unrelated endpoint, the browser will pop-up a cert picker dialog (see screenshot).

This is a very bad user experience.

### Describe the solution you'd like
Similar to the IIS location element, we'd need a way to scope the client cert requirement to a path, to remove the side-effects this has on the rest of the application.

### Additional context
<img width="1506" alt="Screenshot 2019-12-30 10 38 48" src="https://user-images.githubusercontent.com/1454075/71577140-44eca380-2af3-11ea-9e71-7129c1576765.png">

**Comment by Tratcher** [comment]
This is a limitation of SslSteam, it does not support explicit renegotiation.

**Comment by leastprivilege** [comment]
OK. Thanks. Then it would be useful to add docs that describe how to achieve that with IIS and nginx.

**Comment by gynet** [comment]
It seems like Http.sys does not have such issue.  Also if ppl use http.sys, is there any way to enforce ssl protocol to be TSL1.2. Any way to work around for his issue? 

I know Kestrel can enforce TSL1.2 by following, not see any opts for HttpSys
```
   opt.SslProtocols = SslProtocols.Tls12;
```

**Comment by Tratcher** [comment]
@gynet HttpSys TLS protocols can't be configured in the app. You can either use netsh.exe to set disablelegacytls, or disable old TLS versions system wide in the registry.
https://docs.microsoft.com/en-us/security/disable-legacy-tls

**Comment by leastprivilege** [comment]
OK - from my POV this is resolved and can be closed.

Kestrel can't do it. IIS can. But given re-negotiation is not really encouraged, using a sub-domain to isolate client cert requirements seems the way to go.

I have that working with nginx. Let me know if this is interesting for docs.

**Comment by Tratcher** [comment]
@leastprivilege by all means, that would be a good doc addition.

**Comment by gynet** [comment]
Currently, the "Allow Certificate" option is not working as expected as it will prompt user a window to request cert. It would be better to document the behavior.

**Comment by anurse** [comment]
A docs change sounds like a good plan here. Something showing how to have a section of your app that requires client certs (while the rest does not) using separate host bindings (since TLS certificates are negotiated before the path is actually available anyway).

**Comment by gynet** [comment]
Our customers' run-time clients and our UI are share the same service endpoints, and they are using both OAuth and client cert-based auth. It would be very difficult to make such change without breaking anything. Is there any way to workaround the browser prompt?

**Comment by Tratcher** [comment]
No, this is a fundamental TLS/HTTPS issue, not a Kestrel specific problem. Certificates are an artifact of the TLS layer so they can't reliably be made conditional based on any HTTP request information.

**Comment by Tratcher** [comment]
FYI, Kestrel needs an SslStream feature (https://github.com/dotnet/corefx/issues/41622) so it can specify client cert requirements per host name. Otherwise you need to setup your subdomain on a separate IP.

**Comment by Rick-Anderson** [comment]
@leastprivilege 
> I have that working with nginx. Let me know if this is interesting for docs.

Let me know how I can help.

**Comment by Tratcher** [comment]
Some docs were added for this here:
https://docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.1#optional-client-certificates

Cross linking from the servers might help with discovery.
